app/routes/items/detail.js

```diff
@@ -36,20 +45,87 @@ export default AuthorizeRoute.extend({
     var previousRoutes = this.router.router.currentHandlerInfos;
     var previousRoute = previousRoutes && previousRoutes.pop();
     var path = "items.index";
-    if(previousRoute) {
+    if (previousRoute) {
       var routeName = previousRoute.name;
-      if(routeName.indexOf("items") === 0) {
+      if (routeName.indexOf("items") === 0) {
         path = this.get("itemBackLinkPath") || path;
-      } else if(routeName.indexOf("items") > -1 || routeName === "orders.detail"){
+      } else if (
+        routeName.indexOf("items") > -1 ||
+        routeName === "orders.detail"
+      ) {
         path = routeName;
       }
     }
     this.set("itemBackLinkPath", path);
   },
 
-  setupController(controller, model){
+  setupController(controller, model) {
     this._super(controller, model);
-    controller.set('callOrderObserver', false);
-    controller.set('backLinkPath', this.get('itemBackLinkPath'));
+    controller.set("showSetList", false);
+    controller.set("callOrderObserver", false);
+    controller.set("backLinkPath", this.get("itemBackLinkPath"));
+    controller.set("active", true);
+  },
+
+  resetController(controller, isExiting) {
+    if (isExiting) {
+      controller.set("active", false);
+    }
+  },
+
+  // --- Helpers
+
+  /**
+   * Loads all the images of an item (cached)
+   *
+   * @param {Item} item the selected item
+   */
+  preloadImages(item) {
+    const ids = item.getWithDefault("imageIds", []);
+    return Ember.RSVP.all(ids.map(id => this.loadImage(id)));
+  },
+
+  /**
+   * Loads an image if not available
+   *
+   * @param {String} id the image id
+   * @returns {Image} the image record
+   */
+  loadImage(id) {
+    const cachedImg = this.store.peekRecord("image", id);
+    if (cachedImg) {
+      return Ember.RSVP.resolve(cachedImg);
```

Comment:
Do we need `Ember.RSVP.resolve` here? or we can only return `cachedImg` value?

Reply:
Kinda depends on how the caller method uses the returned value.
But we can't make any assumptions, therefore the returned type should always be a promise.

But I will transform this into an `async` method, to clear the `resolve` (same result, better looking)